server/src/test/java/org/elasticsearch/tasks/CancellableTasksTrackerTests.java

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.tasks;
+
+import org.elasticsearch.test.ESTestCase;
+
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.BrokenBarrierException;
+import java.util.concurrent.CyclicBarrier;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static org.hamcrest.Matchers.empty;
+import static org.hamcrest.Matchers.hasItem;
+import static org.hamcrest.Matchers.lessThanOrEqualTo;
+import static org.hamcrest.Matchers.not;
+
+public class CancellableTasksTrackerTests extends ESTestCase {
+
+    private static class TestTask {
+        private final Thread actionThread;
+        private final Thread watchThread;
+        private final Thread concurrentRemoveThread;
+
+        // 0 == before put, 1 == during put, 2 == after put, before remove, 3 == during remove, 4 == after remove
+        private final AtomicInteger state = new AtomicInteger();
+        private final boolean concurrentRemove = randomBoolean();
+
+        TestTask(Task task, String item, CancellableTasksTracker<String> tracker, Runnable awaitStart) {
+            if (concurrentRemove) {
+                concurrentRemoveThread = new Thread(() -> {
+                    awaitStart.run();
+
+                    for (int i = 0; i < 10; i++) {
+                        if (3 <= state.get()) {
+                            final String removed = tracker.remove(task);
+                            if (removed != null) {
+                                assertSame(item, removed);
+                            }
+                        }
+                    }
+                });
+            } else {
+                concurrentRemoveThread = new Thread(awaitStart);
+            }
+
+            actionThread = new Thread(() -> {
+                awaitStart.run();
+
+                state.incrementAndGet();
+                tracker.put(task, item);
+                state.incrementAndGet();
+
+                state.incrementAndGet();
+                final String removed = tracker.remove(task);
+                state.incrementAndGet();
+                if (concurrentRemove == false || removed != null) {
+                    assertSame(item, removed);
+                }
+
+                assertNull(tracker.remove(task));
+            }, "action-thread-" + item);
+
+            watchThread = new Thread(() -> {
+                awaitStart.run();
+
+                for (int i = 0; i < 10; i++) {
```

Comment:
Should this not always run at least until state.get() ==4 (and then another round)?

Reply:
I found that this sometimes noticeably added to the test runtime, we spent ages doing this loop before actually doing the useful work. Again, in practice this covers what we care about.